File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_update.pl
#!/usr/bin/env perl
#     R. Mankel, DESY Hamburg     06-Jul-2007
#     A. Parenti, DESY Hamburg    16-Apr-2008
#     $Revision: 1.7 $ by $Author: jbehr $
#     $Date: 2012/09/10 15:11:05 $
#
#  Update local mps database with batch job status
#  
#
#  Usage:
#

BEGIN {
use File::Basename;
unshift(@INC, dirname($0)."/mpslib");
}
use Mpslib;

read_db();

# initialize FLAG array. Mark jobs we do not have to worry about
my @FLAG = -1;
my $submittedjobs = 0;
for ($i=0; $i<@JOBID; ++$i) {
  if (         $JOBSTATUS[$i] =~ /SETUP/i
	       or $JOBSTATUS[$i] =~ /DONE/i
	       or $JOBSTATUS[$i] =~ /FETCH/i
	       or $JOBSTATUS[$i] =~ /OK/i
	       or $JOBSTATUS[$i] =~ /ABEND/i
	       or $JOBSTATUS[$i] =~ /FAIL/i
     #          or $JOBSTATUS[$i] =~ /DISABLED/i
     )
    {
    $FLAG[$i] = 1; # no need to care
  }
  else {
    $FLAG[$i] = -1; # care!
    $submittedjobs++;
  }
}
print "submitted jobs $submittedjobs\n";
if ( $submittedjobs > 0) {
  my $in = `bjobs -l`;
  #my $in = `cat output`;
  unless($in =~ /No unfinished job found/i)
    {
      $in =~ s/\n//g;
      $in =~ s/\s\s//g;
      $in =~ s/\s+//g;
      my @result = split /\-{50}/, $in;
      foreach my $i (@result) {
        chomp $i;
        #my ($status) = $i =~ /Status<(.+?)>/;
        #my ($jobid) = $i =~ /Job<(.+?)>,/;
        if (my ($status) = $i =~ /Status<([A-Z]+?)>/) {
          if (my ($jobid) = $i =~ /Job<(\d+?)>,/) {
            my $cputime = 0;
            $cputime = $1 if ($i =~ /TheCPUtimeusedis(\d+?)seconds/);
            print "out $status $jobid $cputime\n";
            my $theIndex = -1;
            my $disabled = "";
            for (my $k=0; $k<@JOBID; ++$k) {
              if ($JOBID[$k] == $jobid) {
                $theIndex = $k;
                $disabled = "DISABLED" if ($JOBSTATUS[$k] =~ /DISABLED/i);
              }
              #print "For index $k check jobid $JOBID[$k] result $theIndex\n";
            }
            if ($theIndex == -1) {
              print "mps_update.pl - the job $jobid was not found in the JOBID array\n";
              #exit(-1);
            }
            next if($theIndex == -1);
            next if($FLAG[$theIndex] == 1);
            $JOBSTATUS[$theIndex] = $disabled.$status;
            if ($status eq "RUN" || $status eq "DONE") {
              if ($cputime>0) {
                my $diff = $cputime - $JOBRUNTIME[$theIndex];
                $JOBRUNTIME[$theIndex] = $1;
                $JOBHOST[$theIndex] = "+$diff";
                $JOBINCR[$theIndex] = $diff;
              } else {
                $JOBRUNTIME[$theIndex] = 0;
                $JOBINCR[$theIndex] = 0;
              }
            }
            $FLAG[$theIndex] = 1;
            print "set flag of job $theIndex with id $JOBID[$theIndex] to 1\n";
          }
        }
      }
    }
}

# loop over remaining jobs to see whether they are done
my $theIndex = -1;
for ($i=0; $i<@JOBID; ++$i) {
  $theIndex = $i;
  my $disabled = "";
  $disabled = "DISABLED" if ($JOBSTATUS[$i] =~ /DISABLED/i);
  print " DB job $JOBID[$i] flag $FLAG[$theIndex]\n";
  if ($FLAG[$theIndex] == 1) {
    next;
  }
  # check whether job may be done
  $theBatchDirectory = sprintf "LSFJOB\_%d",$JOBID[$i]; #GF: $theIndex??
  print "theBatchDirectory $theDirectory\n";
  ## if (-d "LSFJOB\_$JOBID[$i]") {
  ##  print "LSFJOB\_$JOBID[$i] exists\n";
  if (-d $theBatchDirectory) {
    print "Directory $theBatchDirectory exists\n";
    $JOBSTATUS[$theIndex] = $disabled."DONE";
  } else {
    if ($JOBSTATUS[$theIndex] =~ /RUN/i) {
      print "WARNING: Job $theIndex in state RUN, neither found by bjobs nor find LSFJOB directory!\n";
      # FIXME: check if job not anymore in batch system
      # might set to FAIL - but probably $theBatchDirectory is just somewhere else...
    }
  }
}


# check for orphaned jobs
for ($i=0; $i<@JOBID; ++$i) {
  unless ($FLAG[$i] eq 1) {
    unless ($JOBSTATUS[$i] =~ /SETUP/i
            or $JOBSTATUS[$i] =~ /DONE/i
            or $JOBSTATUS[$i] =~ /FETCH/i
            or $JOBSTATUS[$i] =~ /TIMEL/i
            or $JOBSTATUS[$i] =~ /SUBTD/i) {
      print "Funny entry index $i job $JOBID[$i] status $JOBSTATUS[$i]\n";
    }
  }
}
write_db();

File: Alignment/MillePedeAlignmentAlgorithm/scripts/mps_update.py
#!/usr/bin/env python3
from __future__ import print_function
from builtins import range
import os
import re
import subprocess
import Alignment.MillePedeAlignmentAlgorithm.mpslib.Mpslibclass as mpslib

import six

def fill_time_info(mps_index, status, cpu_time):
    """Fill timing info in the database for `mps_index`.

    Arguments:
    - `mps_index`: index in the MPS database
    - `status`: job status
    - `cpu_time`: extracted CPU timing information
    """

    cpu_time = int(round(cpu_time))  # care only about seconds for now
    if status in ("RUN", "DONE"):
        if cpu_time > 0:
            diff = cpu_time - lib.JOBRUNTIME[mps_index]
            lib.JOBRUNTIME[mps_index] = cpu_time
            lib.JOBHOST[mps_index] = "+"+str(diff)
            lib.JOBINCR[mps_index] = diff
        else:
            lib.JOBRUNTIME[mps_index] = 0
            lib.JOBINCR[mps_index] = 0



################################################################################
# mapping of HTCondor status codes to MPS status
htcondor_jobstatus = {"1": "PEND", # Idle
                      "2": "RUN",  # Running
                      "3": "EXIT", # Removed
                      "4": "DONE", # Completed
                      "5": "PEND", # Held
                      "6": "RUN",  # Transferring output
                      "7": "PEND"} # Suspended


################################################################################
# collect submitted jobs (use 'in' to handle composites, e.g. DISABLEDFETCH)
lib = mpslib.jobdatabase()
lib.read_db()

submitted_jobs = {}
for i in range(len(lib.JOBID)):
    submitted = True
    for status in ("SETUP", "OK", "DONE", "FETCH", "ABEND", "WARN", "FAIL"):
        if status in lib.JOBSTATUS[i]:
            submitted = False
            break
    if submitted:
        submitted_jobs[lib.JOBID[i]] = i
print("submitted jobs:", len(submitted_jobs))


################################################################################
# deal with submitted jobs by looking into output of shell (condor_q)
if len(submitted_jobs) > 0:
    job_status = {}
    condor_q = subprocess.check_output(["condor_q", "-af:j",
                                        "JobStatus", "RemoteSysCpu"],
                                       stderr = subprocess.STDOUT)
    for line in condor_q.splitlines():
        job_id, status, cpu_time = line.split()
        job_status[job_id] = {"status": htcondor_jobstatus[status],
                              "cpu": float(cpu_time)}

    for job_id, job_info in six.iteritems(job_status):
        mps_index = submitted_jobs.get(job_id, -1)
        # check for disabled Jobs
        disabled = "DISABLED" if "DISABLED" in lib.JOBSTATUS[mps_index] else ""

        # continue with next batch job if not found or not interesting
        if mps_index == -1:
            print("mps_update.py - the job", job_id, end=' ')
            print("was not found in the JOBID array")
            continue
        else:                   # pop entry from submitted jobs
            submitted_jobs.pop(job_id)


        # if found update Joblists for mps.db
        lib.JOBSTATUS[mps_index] = disabled+job_info["status"]
        fill_time_info(mps_index, job_info["status"], job_info["cpu"])


################################################################################
# loop over remaining jobs to see whether they are done
for job_id, mps_index in submitted_jobs.items(): # IMPORTANT to copy here (no iterator!)
    # check if current job is disabled. Print stuff.
    disabled = "DISABLED" if "DISABLED" in lib.JOBSTATUS[mps_index] else ""
    print(" DB job ", job_id, mps_index)

    # check if it is a HTCondor job already moved to "history"
    userlog = os.path.join("jobData", lib.JOBDIR[mps_index], "HTCJOB")
    condor_h = subprocess.check_output(["condor_history", job_id, "-limit", "1",
                                        "-userlog", userlog,
                                        "-af:j", "JobStatus", "RemoteSysCpu"],
                                       stderr = subprocess.STDOUT)
    if len(condor_h.strip()) > 0:
        job_id, status, cpu_time = condor_h.split()
        status = htcondor_jobstatus[status]
        lib.JOBSTATUS[mps_index] = disabled + status
        fill_time_info(mps_index, status, float(cpu_time))
        submitted_jobs.pop(job_id)
        continue

    if "RUN" in lib.JOBSTATUS[mps_index]:
        print("WARNING: Job ", mps_index, end=' ')
        print("in state RUN, neither found by htcondor, nor bjobs, nor find", end=' ')
        print("LSFJOB directory!")


################################################################################
# check for orphaned jobs
for job_id, mps_index in six.iteritems(submitted_jobs):
    for status in ("SETUP", "DONE", "FETCH", "TIMEL", "SUBTD"):
        if status in lib.JOBSTATUS[mps_index]:
            print("Funny entry index", mps_index, " job", lib.JOBID[mps_index], end=' ')
            print(" status", lib.JOBSTATUS[mps_index])


lib.write_db()
